exercise4.go

```diff
@@ -0,0 +1,39 @@
+package main
+
+// Modify dup2 to print the names of all files
+// in which each duplicated line occurs
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+)
+
+func main() {
+	counts := make(map[string]int)
+	files := os.Args[1:]
+	if len(files) == 0 {
+		countLines(os.Stdin, counts, "")
+	} else {
+		for _, arg := range files {
+			f, err := os.Open(arg)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "dup2: %v\n", err)
+				continue
+			}
+			countLines(f, counts, arg)
+			f.Close()
+		}
+	}
+}
+
+func countLines(f *os.File, counts map[string]int, fileName string) {
+	input := bufio.NewScanner(f)
+	for input.Scan() {
+		counts[input.Text()]++
+		if counts[input.Text()] > 1 {
+			fmt.Println(fileName)
+		}
```

Comment:
There is a bug here. This will miss the first file a duplicated line appears in.  Consider your test output:

```
./random.txt
./random.txt
./random.txt
./random2.txt
./random2.txt
```

`random.txt` contains two "inter file" duplicates: `blue` and `yellow` in addition to the three "intra file" duplicates but only shows up 3 times in your ouput instead of 5 times.

To highlight this, please change the output to show something like:

```
hi
    duplicated in ./random.txt
hello
    duplicated in ./random.txt
random
    duplicated in ./random.txt
blue
    duplicated in ./random.txt, ./random2.txt
yellow
    duplicated in ./random.txt, ./random2.txt
```

Ordering of the duplicates or files in the output doesn't matter. I believe this is closer to the intent of the exercise than the current implementation.  If you want to go above & beyond, keep the counts too:

```
blue
     2 times in ./random.txt, 1 times in ./random2.txt
```